packages/perseus-core/src/widgets/matcher/matcher-util.ts

```diff
@@ -0,0 +1,138 @@
+type MatcherInfo = {
+    left: ReadonlyArray<string>;
+    right: ReadonlyArray<string>;
+    orderMatters: boolean;
+    problemNum: number | null | undefined;
+};
+
+type MatcherShuffleInfo = {
+    left: ReadonlyArray<string>;
+    right: ReadonlyArray<string>;
+    orderMatters: boolean;
+};
+
+// Note: Previously called cyrb128
+function getHashes(str: string): [number, number, number, number] {
+    let h1 = 1779033703;
+    let h2 = 3144134277;
+    let h3 = 1013904242;
+    let h4 = 2773480762;
+    let k = 0;
+    for (let i = 0; i < str.length; i++) {
+        k = str.charCodeAt(i);
+        h1 = h2 ^ Math.imul(h1 ^ k, 597399067);
+        h2 = h3 ^ Math.imul(h2 ^ k, 2869860233);
+        h3 = h4 ^ Math.imul(h3 ^ k, 951274213);
+        h4 = h1 ^ Math.imul(h4 ^ k, 2716044179);
+    }
+    h1 = Math.imul(h3 ^ (h1 >>> 18), 597399067);
+    h2 = Math.imul(h4 ^ (h2 >>> 22), 2869860233);
+    h3 = Math.imul(h1 ^ (h3 >>> 17), 951274213);
+    h4 = Math.imul(h2 ^ (h4 >>> 19), 2716044179);
+    return [
+        (h1 ^ h2 ^ h3 ^ h4) >>> 0,
+        (h2 ^ h1) >>> 0,
+        (h3 ^ h1) >>> 0,
+        (h4 ^ h1) >>> 0,
+    ];
+}
+
+// Note: Previously called sfc32
+function getSeedFunctionFromHashes(
+    a: number,
+    b: number,
+    c: number,
+    d: number,
+): () => number {
+    return function (): number {
+        a |= 0;
+        b |= 0;
+        c |= 0;
+        d |= 0;
+        const t = (((a + b) | 0) + d) | 0;
+        d = (d + 1) | 0;
+        a = b ^ (b >>> 9);
+        b = (c + (c << 3)) | 0;
+        c = (c << 21) | (c >>> 11);
+        c = (c + t) | 0;
+        return (t >>> 0) / 4294967296;
+    };
+}
+
+// TODO(LEMS-2841): Should be able to remove once getPublicWidgetOptions is hooked up
+export const matcherShuffle = (
+    props: MatcherInfo,
+): {left: ReadonlyArray<string>; right: ReadonlyArray<string>} => {
+    // Use the same random() function to shuffle both columns sequentially
+    const rng = CoreUtil.seededRNG(props.problemNum as number);
+
+    let left;
+    if (!props.orderMatters) {
+        // If the order doesn't matter, don't shuffle the left column
+        left = props.left;
+    } else {
+        left = CoreUtil.shuffle(props.left, rng, /* ensurePermuted */ true);
+    }
+
+    const right = CoreUtil.shuffle(props.right, rng, /* ensurePermuted */ true);
+
+    return {left, right};
+};
+
+function matcherShuffleFromHash(
+    data: MatcherShuffleInfo,
+    seed: number,
+): {left: ReadonlyArray<string>; right: ReadonlyArray<string>} {
+    // Use the same random() function to shuffle both columns sequentially
+    let left;
+    if (!data.orderMatters) {
+        // If the order doesn't matter, don't shuffle the left column
+        left = data.left;
+    } else {
+        left = CoreUtil.shuffle(data.left, seed, /* ensurePermuted */ true);
+    }
+
+    const right = CoreUtil.shuffle(data.right, seed, /* ensurePermuted */ true);
```

Comment:
@benchristel Is there still a concern with these `ensurePermuted` arguments leaking data? Is that something we can fix?

Reply:
I think we should keep `ensurePermuted = true`. It's necessary to preserve the existing behavior. I think it would be confusing (and maybe even blocking?) to learners if we ever displayed the correct answer as the initial state, since we don't enable the "submit" button until the learner changes the state of every widget.